Add a pattern which detects local variable refactorings

zacikpa opened the pull request:
The pattern works as follows:
- Differing allocas of local memory and stores to such memory are ignored.
- When there is a load from such memory, identifying the replacement works the same as with repetitive loads, i.e., a suitable store has to be found in all backward CFG paths from the load.

Similar to the repetitive loads, this approach is limited by possibly conflicting stores and function calls. Further improvements may be possible using pointer alias analysis (maybe using LLVM's `MemoryDependenceAnalysis` pass?).

In the process of implementing this, I found multiple bugs in the repetitive load skipping implementation; see the first commit.

I'm not really sure about the pattern name; maybe something like `local-memory` would be more suitable?

viktormalik commented on diffkemp/simpll/DifferentialFunctionComparator.cpp:
```diff
@@ -595,99 +616,114 @@ bool DifferentialFunctionComparator::maySkipCast(const User *Cast) const {
+    return std::all_of(Op->user_begin(), Op->user_end(), [&](auto user) {
+        if (auto userBinOp = dyn_cast<BinaryOperator>(user))
+            return userBinOp->getOpcode() == Op->getOpcode();
+        if (auto userStore = dyn_cast<StoreInst>(user))
+            return hasIgnoredAllocaAsPtrOperand(userStore);
```
Why do we need to have this check here? This pattern doesn't seem specific to binary operations, could we come up with some general approach that would allow us to avoid this?

Also this is missing the `config.Patterns.StructChanges` check.

viktormalik replied:
Idea: add a method which will follow the stored value and return uses of all loads from the same pointer. Then, check that all uses of these are binary operations.

zacikpa replied:
I decided not to follow this approach because I think it's unnecessary. If the result of the skipped binary operation is stored to an ignored variable, we know that the store will always be skipped, which should never lead to false negatives.

I did, however, refactor `maySkipInstruction` to make this more general. See my comment summarizing the changes.

viktormalik reviewed (commented):
I'm a bit confused about what situations this pattern tries to cover. The unit test features a replacement of multiple local constants (possibly local variables?) by a single local variable of a struct type. In literature, this is called *group a set of variables into a new structure*, in which case we could call it `group-vars` in DiffKemp.

On the other hand, the pattern seems to be more general than that because it doesn't store a mapping between the original values (constants) and the new memory locations (fields of a local struct) - or I've failed to find that. As I noted in one of the comments, this approach seems too general and prone to false positives.

Am I missing something?

Besides that, I've a couple of minor comments.

viktormalik commented on diffkemp/simpll/DifferentialFunctionComparator.cpp:
```diff
@@ -504,10 +505,22 @@ int DifferentialFunctionComparator::cmpAllocs(const CallInst *CL,
+/// they need a replacement.
 bool DifferentialFunctionComparator::maySkipInstruction(
         const Instruction *Inst) const {
+    if (Inst->isSafeToRemove()
+        && (Inst->getNumUses() == 0
+            || (config.Patterns.StructChanges
+                && isOnlyUsedByIgnorableMemInstructions(Inst)))) {
```
```suggestion
    if (config.Patterns.StructChanges && Inst->isSafeToRemove() && isOnlyUsed...) {
```

viktormalik commented on diffkemp/simpll/DifferentialFunctionComparator.cpp:
```diff
@@ -1379,6 +1415,12 @@ int DifferentialFunctionComparator::cmpFieldAccess(
 
+    // If we previously ignored allocation of memory that any of the two GEPs
+    // point to, fail (cmpValues would compare their pointer operands as equal).
+    if (hasIgnoredAllocaAsPtrOperand(GEPL)
+        || hasIgnoredAllocaAsPtrOperand(GEPR))
+        RETURN_WITH_LOG(1);
```
What is the purpose of this? It can lead to false positives so we double-check why it is necessary.

zacikpa replied:
I didn't figure it out. It might have been relevant before I added some other check elsewhere, but everything seems to work fine without this now.

zacikpa commented on diffkemp/simpll/DifferentialFunctionComparator.cpp:
```diff
@@ -504,10 +505,22 @@ int DifferentialFunctionComparator::cmpAllocs(const CallInst *CL,
-/// inside the ignored instructions map.
+/// semantics). Ignorable instructions are stored inside the replaced
+/// instructions map or the ignored instructions set, depending on whether
+/// they need a replacement.
```
Updated.

zacikpa commented on diffkemp/simpll/DifferentialFunctionComparator.cpp:
```diff
@@ -595,99 +616,114 @@ bool DifferentialFunctionComparator::maySkipCast(const User *Cast) const {
+            }
+            if (ReplacementCandidate) {
+                if (Replacement) {
+                    // If another suitable replacement has been found previously
+                    // in a different CFG path, fail. We need a single one.
```
Yes, exactly.

zacikpa commented on diffkemp/simpll/DifferentialFunctionComparator.cpp:
```diff
@@ -595,99 +616,114 @@ bool DifferentialFunctionComparator::maySkipCast(const User *Cast) const {
-                PreviousLoad = nullptr;
-                break;
+                // Fail if no more predecessors are available.
+                return false;
```
Updated the comment to be more explicit about this, thanks.

zacikpa commented on diffkemp/simpll/DifferentialFunctionComparator.cpp:
```diff
@@ -528,6 +541,14 @@ bool DifferentialFunctionComparator::maySkipInstruction(
+    if (auto Store = dyn_cast<StoreInst>(Inst)) {
+        if (config.Patterns.StructChanges
+            && hasIgnoredAllocaAsPtrOperand(Store)) {
+            // Skip a store if it points to a previously ignored alloca.
```
As we discussed in person, this could not lead to false negatives due to the check at the beginning of `cmpValues` which disabled value comparison of skipped allocas.

zacikpa replied:
In the updated version, where we _can_ "un-ignore" allocas, this is a problem which I fixed by converting the `ignoredInstruction` set to a map. The value in the map is a bool which specifies whether the ignored instruction can be "un-ignored" in the future. If we ever skip a store to the alloca before un-ignoring it, the value is set to false for the alloca, and from that point onward, it cannot be un-ignored.

zacikpa commented on diffkemp/simpll/Utils.cpp:
```diff
@@ -874,3 +875,66 @@ StructType *getTypeByName(const Module &Mod, StringRef Name) {
+/// instruction may store to the memory pointed to by the pointer. This is
+/// currently implemented only as a set of heuristics - if they do not suffice,
+/// we return true.
+bool mayStoreToMemoryAtPtr(const Instruction *Inst, const Value *Ptr) {
```
Fixed, together with other naming issues that we talked about.

zacikpa commented:
Thanks for the thorough review. I updated the PR with the following changes:

- The pattern is now named `group-vars`.
- I refactored a couple of utility functions so as not to duplicate code.
- `ignoredInstructions` is now a map that stores the information on whether the ignored instruction can be "un-ignored" if it is successfully compared (fully) in `cmpValues`. This is currently only used for ignored `alloca` instructions.
- New logic added to the beginning of `cmpValues` related to the above: if we encounter two ignored instructions that can both be un-ignored, we attempt to compare them properly.
- When skipping a `store` to an ignored `alloca`, the `alloca` can't be un-ignored from that point onward.
- Generalized skipping instructions based on their users; in some cases, an instruction may be skipped because it only has certain kinds of users. The `maySkipInstruction` function now collects alternative conditions that the users may fulfill in order to allow instruction skipping.

viktormalik commented:
One high-level comment (I'm still trying to figure out the code):

> Thanks for the thorough review. I updated the PR with the following changes:
> 
> 
> 
> - The pattern is now named `group-vars`.
> 
> - I refactored a couple of utility functions so as not to duplicate code.
> 
> - `ignoredInstructions` is now a map that stores the information on whether the ignored instruction can be "un-ignored" if it is successfully compared (fully) in `cmpValues`. This is currently only used for ignored `alloca` instructions.
> - New logic added to the beginning of `cmpValues` related to the above: if we encounter two ignored instructions that can both be un-ignored, we attempt to compare them properly.

Idea: could we separate "ignored instructions" which are never compared again and "skipped instructions" which are temporarily skipped but compared again once encountered as operands?

> - When skipping a `store` to an ignored `alloca`, the `alloca` can't be un-ignored from that point onward.
> 
> - Generalized skipping instructions based on their users; in some cases, an instruction may be skipped because it only has certain kinds of users. The `maySkipInstruction` function now collects alternative conditions that the users may fulfill in order to allow instruction skipping.

zacikpa commented:
I updated the PR again.

> Idea: could we separate "ignored instructions" which are never compared again and "skipped instructions" which are temporarily skipped but compared again once encountered as operands?

Good idea. It requires updating a couple of checks (newly checking both `ignored` and `skipped` instructions), but I believe it is still worth it. Done.

Let's not support obscure combinations of patterns for now. I reverted all changes to the `reordered-bin-ops` pattern, which left three major changes in this PR. I force-pushed them as new commits (sorry for that;  the changes were starting to get really messy).

@viktormalik 
All the comments from the first review should be resolved.

zacikpa commented:
New edit: Swapped the meaning of skipped and ignored instructions (skipped instructions can be compared again as operands; ignoring is irreversible).

viktormalik commented on diffkemp/simpll/Utils.cpp:
```diff
@@ -874,3 +875,66 @@ StructType *getTypeByName(const Module &Mod, StringRef Name) {
+/// Given an instruction and a pointer value, try to determine whether the
+/// instruction may store to the memory pointed to by the pointer. This is
+/// currently implemented only as a set of heuristics - if they do not suffice,
+/// we return true.
```
Yet there's `return false` at the end of the function :-)

zacikpa replied:
:) Reworded.

viktormalik reviewed (commented):
Overall looks good to me, just some remarks and details for comments and naming.

zacikpa commented on diffkemp/simpll/DifferentialFunctionComparator.cpp:
```diff
@@ -2025,3 +2111,28 @@ bool DifferentialFunctionComparator::checkInverseCondUsers(
 }
+
+/// Check whether the pointer operand of the given instruction (e.g., load)
+/// is a previously ignored alloca or a GEP pointing to one.
```
Reworded to "local memory allocated by an ignored alloca". The `getAllocaOp` retrieves the alloca through a GEP if necessary.

viktormalik reviewed (approved):
Thanks for all the work put into this pattern!

viktormalik commented:
Can you squash the last commit?
